modules/services/openssh.nix

```diff
@@ -0,0 +1,49 @@
+{ config, lib, ... }:
+
+let
+  cfg = config.services.openssh;
+in
+{
+  options = {
+    services.openssh.enable = lib.mkOption {
+      type = lib.types.nullOr lib.types.bool;
+      default = null;
+      description = ''
+        Whether to enable an OpenSSH server.
+
+        The default is null which means let macOS manage the OpenSSH server.
+      '';
+    };
+
+    services.openssh.useAppleServer = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      example = false;
+      description = ''
+        Whether to use Apple's built-in OpenSSH server.
+      '';
+    };
+  };
+
+  config = {
+    assertions = [
+      {
+        assertion = cfg.useAppleServer;
+        message = "Currently only using Apple's built-in OpenSSH server has been implemented.";
+      }
+    ];
+
+    # We don't use `systemsetup -setremotelogin` as it requires Full Disk Access
+    system.activationScripts.launchd.text = lib.mkIf (cfg.enable != null) (if cfg.enable then ''
+      if [[ "$(systemsetup -getremotelogin | sed 's/Remote Login: //')" == "Off" ]]; then
+        launchctl enable system/com.openssh.sshd
+        launchctl bootstrap system /System/Library/LaunchDaemons/ssh.plist
+      fi
+    '' else ''
+      if [[ "$(systemsetup -getremotelogin | sed 's/Remote Login: //')" == "On" ]]; then
+        launchctl bootout system/com.openssh.sshd
+        launchctl disable system/com.openssh.sshd
+      fi
+    '');
+  };
```

Comment:
The logic for when `cfg.enable` is false worries me a little, because it seems like you might think your SSH server is off and you’re totally protected, but there’s actually a race condition between boot and the activation scripts running where the server can be accessed?

Do we know if there are any meaningful differences between what this does and what `systemsetup -setremotelogin` does? It may be better to just require FDA for this module, as it would certainly solve the race condition.

Reply:
There shouldn't be a race condition because the SSH server is disabled and won't get loaded again on reboot

Reply:
Currently we're adding the text to the `launchd` activation script, which is not one of the activation scripts that runs on boot (`etcChecks`, `etc` and `keyboard`) (Unrelated tangent: we'll probably want `users.users.launchd.agents` to allow us to do user activation when they log in as it might have interactive prompts)

When you `bootout` and `disable` the SSH server, it won't get reenabled on reboot, I've tested this on a VM by manually going into the Sharing menu, enabling Remote Login, doing a `darwin-rebuild switch` with `services.openssh.enable = false;`, rebooting and checking that Remote Login is still disabled

Reply:
Ah, so it actually toggles the visible preference in System Settings when you do this? That’s good enough for me then. Strange that they have a back door like this, though; I wouldn’t be surprised if we have to move to the proper command in future.